src/decoder/lattice-faster-decoder.cc

```diff
@@ -837,6 +856,19 @@ void LatticeFasterDecoder::ProcessNonemitting(BaseFloat cutoff) {
   } // while queue not empty
 }
 
+template void LatticeFasterDecoder::ProcessNonemitting<fst::ConstFst<fst::StdArc>>(BaseFloat cutoff);
+template void LatticeFasterDecoder::ProcessNonemitting<fst::VectorFst<fst::StdArc>>(BaseFloat cutoff);
+template void LatticeFasterDecoder::ProcessNonemitting<fst::Fst<fst::StdArc>>(BaseFloat cutoff);
+
+void LatticeFasterDecoder::ProcessNonemittingWrapper(BaseFloat cost_cutoff) {
+  if (typeid(fst_) == typeid(fst::ConstFst<Arc>)) {
```

Comment:
We haven't used typeid elsewhere and I'm concerned there may be subtleties involved or portability problems.  I think it would be better to use fst_.FstType(), which might be "vector" or "const".
After you do that, re-run your test.  I'm surprised you didn't see a speedup for VectorFst.  You may have to run it a couple times to correct for effects of nfs caching and the like.